File: scrapers/florida.py
# Florida Department of Financial Services — License Lookup Scraper
# Source: https://licenseesearch.fldfs.com
#
# Flow:
#   1. GET / to establish session cookies
#   2. POST / with full form payload (name fields filled, rest empty)
#   3. Parse the response HTML for search results table
#   4. Fetch /Licensee/{id} detail page for status (Valid/Invalid)

import logging
import re
from typing import Optional

import aiohttp
from bs4 import BeautifulSoup

from scrapers.base import LicenseResult, StateScraper

logger = logging.getLogger(__name__)

BASE_URL = "https://licenseesearch.fldfs.com"


class FloridaScraper(StateScraper):
    STATE_CODE = "FL"
    STATE_NAME = "Florida"
    LOOKUP_URL = BASE_URL

    def __init__(self):
        self._session: Optional[aiohttp.ClientSession] = None

    async def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                headers={
                    "User-Agent": (
                        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                        "AppleWebKit/537.36 (KHTML, like Gecko) "
                        "Chrome/120.0.0.0 Safari/537.36"
                    ),
                }
            )
        return self._session

    def _build_form_data(
        self,
        first_name: str = "",
        last_name: str = "",
        license_number: str = "",
        npn: str = "",
    ) -> dict:
        """Build the full POST payload matching the FL DOI form exactly."""
        return {
            "IndividualFNameFilter": first_name.strip(),
            "IndividualLNameFilter": last_name.strip(),
            "IndividualMNameFilter": "",
            "EmailAddressBeginContainFilter": "1",
            "EmailFilter": "",
            "FirmNameBeginContainFilter": "1",
            "FirmNameFilter": "",
            "ResidentStatusFilter": "",
            "FLLicenseNoFilter": license_number.strip(),
            "NPNNoFilter": npn.strip(),
            "LicenseStatusFilter": "1",
            "LicenseCategoryFilter": "",
            "LicenseIssueDateFromFilter": "",
            "LicenseIssueDateToFilter": "",
            "OnlyLicWithNoQuApptFilter": "false",
            "BusinessStateFilter": "",
            "BusinessCityFilter": "",
            "BusinessCountyFilter": "",
            "BusinessZipFilter": "",
            "CEDueDtFromFilter": "",
            "CEDueDtToFilter": "",
            "CEHrsNotMetFilter": "false",
            "AppointingEntityTYCLFilter": "",
            "AppointingEntityStatusFilter": "",
            "AppointingEntityStatusDateFromFilter": "",
            "AppointingEntityStatusDateToFilter": "",
            "LicenseeSearchInfo.PagingInfo.SortBy": "Name",
            "LicenseeSearchInfo.PagingInfo.SortDesc": "False",
            "LicenseeSearchInfo.PagingInfo.CurrentPage": "1",
            "AppointingEntityIdFilter": "",
            "AppointingEntityDisplayName": "",
            "TabLLValue": "0",
            "TabCEValue": "0",
            "TabAppValue": "",
            "hdnLApptEntitySearchListUrl": "/Home/GetAppointingEntityListForSearch",
            "hdnLicenseeSearchListUrl": "/Home/GetLicenseeSearchListPartialView",
        }

    async def _search(self, form_data: dict) -> list[LicenseResult]:
        """Execute the two-step search: POST form to /, then parse results."""
        try:
            session = await self._get_session()

            # Step 1: GET / to establish session cookies
            async with session.get(BASE_URL) as resp:
                if resp.status != 200:
                    logger.error(f"FL GET failed: HTTP {resp.status}")
                    return [LicenseResult(error=f"HTTP {resp.status} on GET")]

            # Step 2: POST / with full form payload
            async with session.post(
                BASE_URL + "/",
                data=form_data,
                headers={
                    "Content-Type": "application/x-www-form-urlencoded",
                    "Referer": BASE_URL,
                },
            ) as resp:
                if resp.status != 200:
                    logger.error(f"FL POST failed: HTTP {resp.status}")
                    return [LicenseResult(error=f"HTTP {resp.status} on POST")]
                html = await resp.text()

            return await self._parse_search_results(html)

        except Exception as e:
            logger.error(f"FL search error: {e}")
            return [LicenseResult(error=str(e))]

    async def lookup_by_name(
        self, first_name: str, last_name: str
    ) -> list[LicenseResult]:
        """Search FL DOI by first + last name."""
        logger.info(f"FL lookup: {first_name} {last_name}")
        form_data = self._build_form_data(first_name=first_name, last_name=last_name)
        return await self._search(form_data)

    async def lookup_by_npn(self, npn: str) -> list[LicenseResult]:
        """Search FL DOI by NPN."""
        logger.info(f"FL NPN lookup: {npn}")
        form_data = self._build_form_data(npn=npn)
        return await self._search(form_data)

    async def lookup_by_license_number(
        self, license_number: str
    ) -> list[LicenseResult]:
        """Search FL DOI by FL license number."""
        logger.info(f"FL license# lookup: {license_number}")
        form_data = self._build_form_data(license_number=license_number)
        return await self._search(form_data)

    async def _parse_search_results(self, html: str) -> list[LicenseResult]:
        """Parse the FL DOI search results page and fetch detail pages."""
        soup = BeautifulSoup(html, "html.parser")
        results = []

        # Check for "No Licensee Detail Found"
        if "no licensee" in html.lower() or "no results" in html.lower():
            logger.info("FL: No results found")
            return [LicenseResult(found=False)]

        # Find the results table
        table = soup.find("table", class_="table")
        if not table:
            logger.warning("FL: Could not find results table")
            return [LicenseResult(found=False)]

        rows = table.find("tbody")
        if not rows:
            return [LicenseResult(found=False)]

        for row in rows.find_all("tr"):
            cells = row.find_all("td")
            if len(cells) < 2:
                continue

            # Column order: Name | License Number | Business Address | City/State/Zip | Email
            name_cell = cells[0]
            license_cell = cells[1] if len(cells) > 1 else None

            # Get the detail page link and name
            link = name_cell.find("a", href=True)
            if not link:
                continue

            full_name = link.get_text(strip=True)
            detail_path = link["href"]  # e.g., /Licensee/1432239
            license_number = license_cell.get_text(strip=True) if license_cell else ""

            # Fetch the detail page to get status, type, expiration
            detail = await self._fetch_detail(detail_path)

            result = LicenseResult(
                found=True,
                state="FL",
                full_name=full_name,
                license_number=license_number,
                status=detail.get("status", "Unknown"),
                active=detail.get("status", "").upper() == "VALID",
                license_type=detail.get("type", ""),
                npn=detail.get("npn", ""),
                expiration_date=detail.get("expiration", ""),
                issue_date=detail.get("issue_date", ""),
                raw_data=detail,
            )
            results.append(result)

            # Only fetch details for first 5 results to avoid rate limiting
            if len(results) >= 5:
                break

        if not results:
            return [LicenseResult(found=False)]

        logger.info(f"FL: Found {len(results)} results")
        return results

    async def _fetch_detail(self, detail_path: str) -> dict:
        """Fetch the licensee detail page and extract status, type, NPN, etc.

        The FL DOI detail page layout:
          - Top section: form-group divs with label + sibling div for values
            (License #, Full Name, Business Address, Email, Phone, County, NPN #)
          - "Valid Licenses" panel: table with Type | Issue Date | Qualifying Appointment
          - "Invalid Licenses" panel: same table structure (or "No invalid licenses found")
          - "Active Appointments" / "Inactive Appointments" panels
        Status is determined by WHICH panel the license appears in (Valid vs Invalid).
        """
        try:
            session = await self._get_session()
            url = BASE_URL + detail_path

            async with session.get(url) as resp:
                if resp.status != 200:
                    return {"error": f"HTTP {resp.status}"}
                html = await resp.text()

            soup = BeautifulSoup(html, "html.parser")

            # ── Extract label: value pairs from form-group divs ──
            # Structure: <div class="form-group ...">
            #              <label class="... control-label">Label:</label>
            #              <div class="col-md-8">Value</div>
            #            </div>
            detail = {}
            for fg in soup.find_all("div", class_="form-group"):
                label_tag = fg.find("label")
                if not label_tag:
                    continue
                label_text = label_tag.get_text(strip=True).rstrip(":")
                # Value div is the sibling of the label inside the same form-group
                value_div = label_tag.find_next_sibling("div")
                if value_div:
                    value = value_div.get_text(" ", strip=True)
                    if value:
                        key = label_text.lower().replace(" ", "_").replace("#", "num")
                        detail[key] = value

            result = {
                "full_name": detail.get("full_name", ""),
                "license_number": detail.get("license_num", detail.get("license", "")),
                "npn": detail.get("npn_num", ""),
                "email": detail.get("email", ""),
                "phone": detail.get("phone", ""),
                "business_address": detail.get("business_address", ""),
                "mailing_address": detail.get("mailing_address", ""),
                "county": detail.get("county", ""),
            }

            # ── Extract license types from Valid / Invalid panels ──
            # Each panel has a panel-heading ("Valid Licenses" / "Invalid Licenses")
            # followed by a table with columns: Type | Issue Date | Qualifying Appointment
            valid_licenses = []
            invalid_licenses = []

            for panel in soup.find_all("div", class_="panel"):
                heading = panel.find("div", class_="panel-heading")
                if not heading:
                    continue
                heading_text = heading.get_text(strip=True).lower()

                table = panel.find("table")
                if not table:
                    continue
                tbody = table.find("tbody")
                if not tbody:
                    continue

                for row in tbody.find_all("tr"):
                    cells = row.find_all("td")
                    if len(cells) >= 2:
                        lic_type = cells[0].get_text(strip=True)
                        issue_date = cells[1].get_text(strip=True)
                        entry = {"type": lic_type, "issue_date": issue_date}

                        if "valid license" in heading_text:
                            entry["status"] = "VALID"
                            valid_licenses.append(entry)
                        elif "invalid license" in heading_text:
                            entry["status"] = "INVALID"
                            invalid_licenses.append(entry)

            # Pick the best license to report:
            # Prefer life/health insurance from valid licenses, then any valid, then invalid
            all_licenses = valid_licenses + invalid_licenses
            chosen = None
            for lic in valid_licenses:
                if "life" in lic["type"].lower() or "health" in lic["type"].lower():
                    chosen = lic
                    break
            if not chosen and valid_licenses:
                chosen = valid_licenses[0]
            if not chosen and invalid_licenses:
                chosen = invalid_licenses[0]

            if chosen:
                result["type"] = chosen["type"]
                result["issue_date"] = chosen.get("issue_date", "")
                result["status"] = chosen["status"]

            # ── Extract expiration from Active/Inactive Appointments ──
            for panel in soup.find_all("div", class_="panel"):
                heading = panel.find("div", class_="panel-heading")
                if not heading:
                    continue
                heading_text = heading.get_text(strip=True).lower()
                if "appointment" not in heading_text:
                    continue

                table = panel.find("table")
                if not table:
                    continue
                tbody = table.find("tbody")
                if not tbody:
                    continue

                for row in tbody.find_all("tr"):
                    cells = row.find_all("td")
                    # Active Appointments: Company Name | Issue Date | Exp Date | Status Date
                    if len(cells) >= 3:
                        exp_date = cells[2].get_text(strip=True)
                        if exp_date and not result.get("expiration"):
                            result["expiration"] = exp_date

            return result

        except Exception as e:
            logger.error(f"FL detail fetch error: {e}")
            return {"error": str(e)}

    async def close(self) -> None:
        if self._session and not self._session.closed:
            await self._session.close()

File: scrapers/florida.go
package scrapers

import (
	"context"
	"fmt"
	"io"
	"log"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
	"github.com/PuerkitoBio/goquery"
)

const flBaseURL = "https://licenseesearch.fldfs.com"

// FloridaScraper scrapes the Florida Department of Financial Services license search.
type FloridaScraper struct {
	sessionFactory SessionFactory
}

// NewFloridaScraper creates a new Florida DOI scraper.
func NewFloridaScraper(sessionFactory SessionFactory) *FloridaScraper {
	return &FloridaScraper{sessionFactory: sessionFactory}
}

// StateCode returns "FL".
func (s *FloridaScraper) StateCode() string { return "FL" }

// ManualLookupURL returns the FL DOI search URL.
func (s *FloridaScraper) ManualLookupURL() string { return flBaseURL }

// LookupByName searches FL DOI by first + last name.
func (s *FloridaScraper) LookupByName(ctx context.Context, firstName, lastName string) ([]LicenseResult, error) {
	log.Printf("FL lookup: %s %s", firstName, lastName)
	form := s.buildFormData(firstName, lastName, "", "")
	return s.search(ctx, form)
}

// LookupByNPN searches FL DOI by National Producer Number.
func (s *FloridaScraper) LookupByNPN(ctx context.Context, npn string) ([]LicenseResult, error) {
	log.Printf("FL NPN lookup: %s", npn)
	form := s.buildFormData("", "", "", npn)
	return s.search(ctx, form)
}

// LookupByLicenseNumber searches FL DOI by FL license number.
func (s *FloridaScraper) LookupByLicenseNumber(ctx context.Context, licenseNumber string) ([]LicenseResult, error) {
	log.Printf("FL license# lookup: %s", licenseNumber)
	form := s.buildFormData("", "", licenseNumber, "")
	return s.search(ctx, form)
}

// buildFormData builds the full POST payload matching the FL DOI form exactly.
func (s *FloridaScraper) buildFormData(firstName, lastName, licenseNumber, npn string) url.Values {
	return url.Values{
		"IndividualFNameFilter":                          {strings.TrimSpace(firstName)},
		"IndividualLNameFilter":                          {strings.TrimSpace(lastName)},
		"IndividualMNameFilter":                          {""},
		"EmailAddressBeginContainFilter":                 {"1"},
		"EmailFilter":                                    {""},
		"FirmNameBeginContainFilter":                     {"1"},
		"FirmNameFilter":                                 {""},
		"ResidentStatusFilter":                           {""},
		"FLLicenseNoFilter":                              {strings.TrimSpace(licenseNumber)},
		"NPNNoFilter":                                    {strings.TrimSpace(npn)},
		"LicenseStatusFilter":                            {"1"},
		"LicenseCategoryFilter":                          {""},
		"LicenseIssueDateFromFilter":                     {""},
		"LicenseIssueDateToFilter":                       {""},
		"OnlyLicWithNoQuApptFilter":                      {"false"},
		"BusinessStateFilter":                            {""},
		"BusinessCityFilter":                             {""},
		"BusinessCountyFilter":                           {""},
		"BusinessZipFilter":                              {""},
		"CEDueDtFromFilter":                              {""},
		"CEDueDtToFilter":                                {""},
		"CEHrsNotMetFilter":                              {"false"},
		"AppointingEntityTYCLFilter":                     {""},
		"AppointingEntityStatusFilter":                   {""},
		"AppointingEntityStatusDateFromFilter":            {""},
		"AppointingEntityStatusDateToFilter":              {""},
		"LicenseeSearchInfo.PagingInfo.SortBy":           {"Name"},
		"LicenseeSearchInfo.PagingInfo.SortDesc":         {"False"},
		"LicenseeSearchInfo.PagingInfo.CurrentPage":      {"1"},
		"AppointingEntityIdFilter":                       {""},
		"AppointingEntityDisplayName":                    {""},
		"TabLLValue":                                     {"0"},
		"TabCEValue":                                     {"0"},
		"TabAppValue":                                    {""},
		"hdnLApptEntitySearchListUrl":                    {"/Home/GetAppointingEntityListForSearch"},
		"hdnLicenseeSearchListUrl":                       {"/Home/GetLicenseeSearchListPartialView"},
	}
}

// search executes the two-step search: GET to establish cookies, then POST form to /.
func (s *FloridaScraper) search(ctx context.Context, form url.Values) ([]LicenseResult, error) {
	session, err := s.sessionFactory()
	if err != nil {
		return nil, fmt.Errorf("fl: session error: %w", err)
	}

	// Step 1: GET / to establish session cookies
	getReq, err := http.NewRequestWithContext(ctx, http.MethodGet, flBaseURL, nil)
	if err != nil {
		return nil, fmt.Errorf("fl: GET request build error: %w", err)
	}
	getReq.Header = http.Header{
		"Accept":            {"text/html"},
		"Accept-Language":   {"en-US,en;q=0.9"},
		http.HeaderOrderKey: {"Accept", "Accept-Language"},
	}

	getResp, err := session.Do(getReq)
	if err != nil {
		return nil, fmt.Errorf("fl: GET request failed: %w", err)
	}
	io.Copy(io.Discard, getResp.Body)
	getResp.Body.Close()

	if getResp.StatusCode != 200 {
		return []LicenseResult{{Error: fmt.Sprintf("HTTP %d on GET", getResp.StatusCode)}}, nil
	}

	// Step 2: POST / with full form payload
	postReq, err := http.NewRequestWithContext(ctx, http.MethodPost, flBaseURL+"/", strings.NewReader(form.Encode()))
	if err != nil {
		return nil, fmt.Errorf("fl: POST request build error: %w", err)
	}
	postReq.Header = http.Header{
		"Content-Type":      {"application/x-www-form-urlencoded"},
		"Referer":           {flBaseURL},
		"Accept":            {"text/html"},
		"Accept-Language":   {"en-US,en;q=0.9"},
		http.HeaderOrderKey: {"Content-Type", "Referer", "Accept", "Accept-Language"},
	}

	postResp, err := session.Do(postReq)
	if err != nil {
		return nil, fmt.Errorf("fl: POST request failed: %w", err)
	}
	defer postResp.Body.Close()

	if postResp.StatusCode != 200 {
		return []LicenseResult{{Error: fmt.Sprintf("HTTP %d on POST", postResp.StatusCode)}}, nil
	}

	// Read the full body so we can check for "no results" text and also parse HTML
	bodyBytes, err := io.ReadAll(postResp.Body)
	if err != nil {
		return nil, fmt.Errorf("fl: read POST body failed: %w", err)
	}
	html := string(bodyBytes)

	return s.parseSearchResults(ctx, session, html)
}

// parseSearchResults parses the FL DOI search results page and fetches detail pages.
func (s *FloridaScraper) parseSearchResults(ctx context.Context, session interface{ Do(req *http.Request) (*http.Response, error) }, html string) ([]LicenseResult, error) {
	lowerHTML := strings.ToLower(html)
	if strings.Contains(lowerHTML, "no licensee") || strings.Contains(lowerHTML, "no results") {
		log.Println("FL: No results found")
		return []LicenseResult{{Found: false, State: "FL"}}, nil
	}

	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return nil, fmt.Errorf("fl: parse HTML error: %w", err)
	}

	table := doc.Find("table.table")
	if table.Length() == 0 {
		log.Println("FL: Could not find results table")
		return []LicenseResult{{Found: false, State: "FL"}}, nil
	}

	tbody := table.First().Find("tbody")
	if tbody.Length() == 0 {
		return []LicenseResult{{Found: false, State: "FL"}}, nil
	}

	var results []LicenseResult
	tbody.Find("tr").Each(func(i int, row *goquery.Selection) {
		if len(results) >= 5 {
			return
		}

		cells := row.Find("td")
		if cells.Length() < 2 {
			return
		}

		nameCell := cells.Eq(0)
		licenseCell := cells.Eq(1)

		link := nameCell.Find("a")
		if link.Length() == 0 {
			return
		}

		fullName := strings.TrimSpace(link.Text())
		detailPath, exists := link.Attr("href")
		if !exists {
			return
		}
		licenseNumber := strings.TrimSpace(licenseCell.Text())

		// Fetch the detail page for status, type, expiration
		detail := s.fetchDetail(ctx, session, detailPath)

		status := detail["status"]
		result := LicenseResult{
			Found:           true,
			State:           "FL",
			FullName:        fullName,
			LicenseNumber:   licenseNumber,
			Status:          status,
			Active:          strings.EqualFold(status, "VALID"),
			LicenseType:     detail["type"],
			NPN:             detail["npn"],
			ExpirationDate:  detail["expiration"],
			IssueDate:       detail["issue_date"],
			BusinessAddress: detail["business_address"],
			BusinessPhone:   detail["phone"],
			Email:           detail["email"],
			County:          detail["county"],
		}
		results = append(results, result)
	})

	if len(results) == 0 {
		return []LicenseResult{{Found: false, State: "FL"}}, nil
	}

	log.Printf("FL: Found %d results", len(results))
	return results, nil
}

// fetchDetail fetches a licensee detail page and extracts status, type, NPN, etc.
//
// The FL DOI detail page layout:
//   - Top section: form-group divs with label + sibling div for values
//     (License #, Full Name, Business Address, Email, Phone, County, NPN #)
//   - "Valid Licenses" panel: table with Type | Issue Date | Qualifying Appointment
//   - "Invalid Licenses" panel: same table structure
//   - "Active Appointments" / "Inactive Appointments" panels
//
// Status is determined by WHICH panel the license appears in (Valid vs Invalid).
func (s *FloridaScraper) fetchDetail(ctx context.Context, session interface{ Do(req *http.Request) (*http.Response, error) }, detailPath string) map[string]string {
	result := map[string]string{}

	detailURL := flBaseURL + detailPath
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, detailURL, nil)
	if err != nil {
		log.Printf("FL detail request build error: %v", err)
		result["error"] = err.Error()
		return result
	}
	req.Header = http.Header{
		"Accept":            {"text/html"},
		"Referer":           {flBaseURL + "/"},
		"Accept-Language":   {"en-US,en;q=0.9"},
		http.HeaderOrderKey: {"Accept", "Referer", "Accept-Language"},
	}

	resp, err := session.Do(req)
	if err != nil {
		log.Printf("FL detail request failed: %v", err)
		result["error"] = err.Error()
		return result
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		result["error"] = fmt.Sprintf("HTTP %d", resp.StatusCode)
		return result
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		log.Printf("FL detail parse error: %v", err)
		result["error"] = err.Error()
		return result
	}

	// ── Extract label:value pairs from form-group divs ──
	// Structure:
	//   <div class="form-group">
	//     <label class="control-label">NPN #:</label>
	//     <div class="col-md-8">12345</div>
	//   </div>
	formFields := map[string]string{}
	doc.Find("div.form-group").Each(func(_ int, fg *goquery.Selection) {
		labelTag := fg.Find("label")
		if labelTag.Length() == 0 {
			return
		}
		labelText := strings.TrimSpace(labelTag.Text())
		labelText = strings.TrimRight(labelText, ":")

		// Value div is the sibling of the label inside the same form-group
		valueDiv := labelTag.Next()
		if valueDiv.Length() == 0 {
			return
		}
		value := strings.TrimSpace(valueDiv.Text())
		if value == "" {
			return
		}

		key := strings.ToLower(labelText)
		key = strings.ReplaceAll(key, " ", "_")
		key = strings.ReplaceAll(key, "#", "num")
		formFields[key] = value
	})

	result["full_name"] = formFields["full_name"]
	if v, ok := formFields["license_num"]; ok {
		result["license_number"] = v
	} else {
		result["license_number"] = formFields["license"]
	}
	result["npn"] = formFields["npn_num"]
	result["email"] = formFields["email"]
	result["phone"] = formFields["phone"]
	result["business_address"] = formFields["business_address"]
	result["mailing_address"] = formFields["mailing_address"]
	result["county"] = formFields["county"]

	// ── Extract license types from Valid / Invalid panels ──
	type licenseEntry struct {
		licType   string
		issueDate string
		status    string
	}
	var validLicenses []licenseEntry
	var invalidLicenses []licenseEntry

	doc.Find("div.panel").Each(func(_ int, panel *goquery.Selection) {
		heading := panel.Find("div.panel-heading")
		if heading.Length() == 0 {
			return
		}
		headingText := strings.ToLower(strings.TrimSpace(heading.Text()))

		table := panel.Find("table")
		if table.Length() == 0 {
			return
		}
		tbody := table.Find("tbody")
		if tbody.Length() == 0 {
			return
		}

		tbody.Find("tr").Each(func(_ int, row *goquery.Selection) {
			cells := row.Find("td")
			if cells.Length() < 2 {
				return
			}
			licType := strings.TrimSpace(cells.Eq(0).Text())
			issueDate := strings.TrimSpace(cells.Eq(1).Text())
			entry := licenseEntry{licType: licType, issueDate: issueDate}

			if strings.Contains(headingText, "valid license") {
				entry.status = "VALID"
				validLicenses = append(validLicenses, entry)
			} else if strings.Contains(headingText, "invalid license") {
				entry.status = "INVALID"
				invalidLicenses = append(invalidLicenses, entry)
			}
		})
	})

	// Pick the best license to report:
	// Prefer life/health insurance from valid licenses, then any valid, then invalid
	var chosen *licenseEntry
	for i := range validLicenses {
		lower := strings.ToLower(validLicenses[i].licType)
		if strings.Contains(lower, "life") || strings.Contains(lower, "health") {
			chosen = &validLicenses[i]
			break
		}
	}
	if chosen == nil && len(validLicenses) > 0 {
		chosen = &validLicenses[0]
	}
	if chosen == nil && len(invalidLicenses) > 0 {
		chosen = &invalidLicenses[0]
	}

	if chosen != nil {
		result["type"] = chosen.licType
		result["issue_date"] = chosen.issueDate
		result["status"] = chosen.status
	}

	// ── Extract expiration from Active/Inactive Appointments ──
	doc.Find("div.panel").Each(func(_ int, panel *goquery.Selection) {
		heading := panel.Find("div.panel-heading")
		if heading.Length() == 0 {
			return
		}
		headingText := strings.ToLower(strings.TrimSpace(heading.Text()))
		if !strings.Contains(headingText, "appointment") {
			return
		}

		table := panel.Find("table")
		if table.Length() == 0 {
			return
		}
		tbody := table.Find("tbody")
		if tbody.Length() == 0 {
			return
		}

		tbody.Find("tr").Each(func(_ int, row *goquery.Selection) {
			if result["expiration"] != "" {
				return
			}
			cells := row.Find("td")
			// Active Appointments: Company Name | Issue Date | Exp Date | Status Date
			if cells.Length() >= 3 {
				expDate := strings.TrimSpace(cells.Eq(2).Text())
				if expDate != "" {
					result["expiration"] = expDate
				}
			}
		})
	})

	return result
}
